XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,540 @@
+free. The destructor will usually run when explicitly called (the case for
+handles) or when a request has completed. In several cases this will be
+triggered by GC. In the case of shared or cached resources, such as
+`HTTPParser`, `destroy()` will be called manually when the TCP connection is no
+longer in need of it. Every subsequent use of a shared resource will have a new
+unique id.
+
+
+## Embedder API
+
+Library developers that handle their own I/O will need to hook into the
+`AsyncWrap` API so that all the appropriate callbacks are called. To
+accommodate this both a C++ and JS API is provided.
+
+
+### Native API
+
+```cpp
+// Helper class users can inherit from, but is not necessary. It will
+// automatically call all four callbacks.
+class AsyncHook {
+  public:
+    AsyncHook(Local<Object> handle, const char* name);
+    ~AsyncHook();
+    MaybeLocal<Value> MakeCallback(
+        const Local<Function> callback,
+        int argc,
+        Local<Value>* argv);
+    Local<Object> handle();
+    uint64_t get_uid();
+  private:
+    AsyncHook();
+    Persistent<Object> handle_;
+    const char* name_;
+    uint64_t uid_;
+}
+
+// Returns the id of the current execution context. If the return value is
+// zero then no execution has been set. This will happen if the user handles
+// I/O from native code.
+uint64_t node::GetCurrentId();
+
+// If the native API doesn't inherit from the helper class then the callbacks
+// must be triggered manually. This triggers the init() callback. The return
+// value is the uid assigned to the handle.
+// TODO(trevnorris): This always needs to be called so that the handle can be
+// placed on the Map for future query.
+uint64_t node::EmitAsyncHandleInit(Local<Object> handle);
+
+// Emit the destroy() callback.
+void node::EmitAsyncHandleDestroy(uint64_t id);
+
+// An API specific to emit before/after callbacks is unnecessary because
+// MakeCallback will automatically call them for you.
+MaybeLocal<Value> node::MakeCallback(Isolate* isolate,
+                                     Local<Object> recv,
+                                     Local<Function> callback,
+                                     uint64_t id,
+                                     int argc,
+                                     Local<Value>* argv);
+```
+
+
+### JS API
+
+There is also a JS API to allow for conceptual correctness. For example, if a
+request batches calls under the hood this would still allow for each request
+to trigger the callbacks individually so that the callback associated with
+each bach request execute inside the correct parent.
+
+This cannot be enforced in any way. It is up to the implementer to make sure
+all of the callbacks are placed and called at the correct time.
+
+```js
+// Returns a new uid incremented from an internal global field.
+async_wrap.newUid();
+
+// Here require the "handle" object is passed so it can be maintained in the
+// Map of handles.
+async_wrap.emitInit(id, handle);
```

Comment:
Would it not be better to return the id like in the `C++` interface? Otherwise it will be possible to call the `init` hook with the same id multiple times.